### PR TITLE
Adds TCG Decks in order to play with them like a TCG.

### PR DESCRIPTION
**Pull request by ArcaneMusic**
## About The Pull Request

Kept you waiting, hunh?
This adds a new multi-faceted object to the game, the TCG Card Pile. Using any 2 cards together on each other will create a pile, and the first card in the pile will determine if the card pile is face-up, or face-down. Depending on the orientation of the card pile, you'll draw cards out of it in the orientation. So, a "deck" card pile would be flipped face down, and would allow you to draw cards out of it also face down, and vis versa.
Decks can be renamed with a pen, and use a radial menu in order to enable quick action on the Trading Card game battlefield.
image
* Draw allows you to draw a single card from the deck, again based on the orientation of the cards used to build the deck.
* Shuffle calls the shuffle() proc on the deck, and (should) randomize the order of the cards in the deck for use.
* Pick up puts the deck into your hands instead of the nonsense dragging required with regular playing cards.
* Flip flips the orientation of the cards in the deck, turning a face up pile into a facedown pile, and a facedown, faceup. Also, calls reverseRange() on the contents of the deck, which should put the bottom of the deck on top, and vis verse, ALSO enabling you to put cards on the bottom or the top of your deck easily.

Card piles also operate with a custom storage component, so you can search for cards in your deck if need be, but also handle edge cases like when you only have 1 card in a pile left, or when you dump out a deck of cards like you would a bag.

## Why It's Good For The Game
 
This basically adds in more sane hands, more sane decks and discard piles, and a massive QOL update to the trading card game to make it actually playable (once potato actually puts the card stats and effects on the cards themselves.)

## Changelog
:cl:
add: By using a TCG card on another TCG card, you can create piles of cards in order to play against each other.
add: Using a deck if tcg cards on a table allows you to shuffle, flip, and draw from the deck as well.
/:cl:

### CODE REVIEW

File: code/game/objects/items/tcg/tcg.dm [hunk 1]
```diff
@@ -71,6 +71,121 @@ GLOBAL_LIST_EMPTY(cached_cards)
...
+	user.visible_message("<span class='notice'>[user] draws a card from \the [src]!</span>", \
+					"<span class='notice'>You draw a card from \the [src]!</span>")
+	if(contents.len == 1)
+		user.transferItemToLoc(contents[contents.len], drop_location())
+		qdel(src)
+
+/obj/item/tcgcard_deck/proc/shuffle_deck(mob/user)
+	if(!contents)
+		return
+	contents = shuffle(contents)
```
Comment (ArcaneMusic) [inline code/game/objects/items/tcg/tcg.dm:185]: Neither shuffle(contents) nor shuttle_inplace(contents) seems to properly shuffle the contents of a deck, not entirely sure why.

File: code/game/objects/items/tcg/tcg.dm [hunk 2]
```diff
@@ -71,6 +71,121 @@ GLOBAL_LIST_EMPTY(cached_cards)
...
+	switch(choice)
+		if("Draw")
+			draw_card(user)
+		if("Shuffle")
+			shuffle_deck(user)
+		if("Pickup")
+			user.put_in_hands(src)
+		if("Flip")
+			flipped = !flipped
+			contents = reverseRange(contents)
```
Comment (ArcaneMusic) [inline code/game/objects/items/tcg/tcg.dm:146]: Similar to the other comment, this one also seems to have an issue when being called on the contents of the object.

File: code/game/objects/items/tcg/tcg.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/game/objects/items/tcg/tcg.dm:185]: shuffle(contents) appears to function properly, can you elaborate?

File: code/game/objects/items/tcg/tcg.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/game/objects/items/tcg/tcg.dm:146]: image
aight so fun fact, shuffle(), unlike reverseRange(), Operates on a Copy() of the list that gets passed to it, which allows it to bypass the fucking restriction on contents. you just need to throw a .Copy() on the end so reverseRange gets a mutable proc for it to play with.

**Comment by LemonInTheDark** [comment]
Is this intended to act as a "hand"? and if so, should you be able to open the radial menu inhand?

File: code/game/objects/items/tcg/tcg.dm [hunk 3]
```diff
@@ -71,6 +71,121 @@ GLOBAL_LIST_EMPTY(cached_cards)
...
+		if(contents.len >= 30)
+			to_chat(user, "<span class='notice'>This pile has too many cards for a regular deck!</span>")
+			return FALSE
+		var/obj/item/tcgcard/new_card = I
+		new_card.flipped = flipped
+		user.transferItemToLoc(new_card, src)
+
+/obj/item/tcgcard_deck/proc/draw_card(mob/user)
+	if(!contents.len)
+		CRASH("A TCG deck was created with no cards inside of it.")
```
Comment (LemonInTheDark) [inline code/game/objects/items/tcg/tcg.dm:170]: CRASH kills the current proc, think of it like a runtime. If you want to continue execution after it runs, use stack_trace()

**Comment by Tharcoonvagh** [comment]
I don't like how there's no way to see the actual cards in the deck, but it's a neat feature

**Comment by LemonInTheDark** [comment]
Ehhhhhhh

Flipping is still broken
Alt clicking a pile, taking out a card, and hitting the pile with that card causes the below
This pile has 4 cards in it but only 3 show, and some are small
image
The contents of the pile do not update if the order gets changed, say by shuffling them
(feature request) I can't hit pile A with pile B to merge them

File: code/datums/components/storage/concrete/trading_cards.dm [hunk 4]
```diff
@@ -54,3 +42,15 @@
...
+/datum/component/storage/concrete/tcg/proc/handle_empty_deck()
+	var/list/contents = contents()
+	//You can't have a deck of one card!
+	if(contents.len <= 1)
+		var/obj/item/tcgcard_deck/deck = parent
+		var/obj/item/tcgcard/card = contents[1]
+		card.forceMove(card.drop_location())
+		card.flipped = deck.flipped
+		card.update_icon_state()
+		card.zoom_in()
```
Comment (LemonInTheDark) [inline code/datums/components/storage/concrete/trading_cards.dm:55]: Couldn't this happen when the deck/card is in my inventory? why would it drop on the ground?

**Review by Twaticus — CHANGES_REQUESTED** [review]
these are weirdly inconsistent regarding text size or backround outline size

**Review by Twaticus — APPROVED** [review]
maybe i just got the 'tisms

File: code/datums/components/storage/concrete/trading_cards.dm [hunk 5]
```diff
@@ -0,0 +1,68 @@
+/**
+  *A storage component to be used on card piles, for use as hands/decks/discard piles. Don't use on something that's not a card pile!
+  */
+/datum/component/storage/concrete/tcg
+	display_numerical_stacking = FALSE
+	max_w_class = WEIGHT_CLASS_TINY
+	max_items = 30
+	max_combined_w_class = WEIGHT_CLASS_TINY * 30
+	///The deck that the card pile is using for FAIR PLAY.
+	var/obj/item/tcgcard_deck/parent_deck
```
Comment (LemonInTheDark) [inline code/datums/components/storage/concrete/trading_cards.dm:10]: this is the same as parent right?

**Pull request merged**